File: scala/binary/src/main/scala/Binary.scala
import Binary._

class Binary(input: BinaryInput) {
  lazy val toDecimal: BinaryOutput =
    parse(input) getOrElse InvalidInputResult
}

object Binary {
  type BinaryInput = String
  type BinaryOutput = Int

  def apply(input: BinaryInput) = new Binary(input)

  val InvalidInputResult: BinaryOutput = 0

  private def parse(input: BinaryInput): Option[BinaryOutput] = {
    def isValidInput(in: BinaryInput) =
      in.nonEmpty && (in forall (c => c == '0' || c == '1'))

    def toDecimal(in: BinaryInput): BinaryOutput =
      in.foldLeft(0) {
        case (output, char) => output * 2 + char.asDigit
      }
//      in.reverse.foldRight(0) {
//        case (char, output) => output * 2 + char.asDigit
//      }

    Option(input) filter isValidInput map toDecimal
  }
}

File: java/binary/src/main/java/Binary.java
import java.util.Optional;
import java.util.function.Function;
import java.util.function.IntBinaryOperator;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class Binary {

    private final String input;

    public Binary(String input) {
        this.input = input;}

    public int getDecimal() {
        return parse(input).orElse(0);
    }

    private static Optional<Integer> parse(String binary) {
        return Optional.ofNullable(binary).filter(isValidBinary).map(toDecimal);
    }

    private static IntBinaryOperator nextDigit =
        (output, bin) -> output * 2 + Character.digit(bin, 10);

    private static Function<String,Integer> toDecimal =
        binary -> binary.chars().reduce(0, nextDigit);

    private static Predicate<String> isValidBinary =
        candidate ->
            !candidate.isEmpty() &&
            characterStream(candidate).allMatch(c -> c == '0' || c == '1');

    private static Stream<Character> characterStream(String str) {
        return str.chars().mapToObj(i -> (char)i);
    }
}
